pinot-connectors/pinot-flink-connector/src/main/java/org/apache/pinot/connector/flink/sink/PinotSinkFunction.java

```diff
@@ -117,31 +123,22 @@ public void close()
   @Override
   public void invoke(T value, Context context)
       throws Exception {
-    _segmentWriter.collect(_recordConverter.convertToRow(value));
+    GenericRow row = _recordConverter.convertToRow(value);
+    _pendingRows.add(row);
+    _segmentWriter.collect(row);
     _segmentNumRecord++;
-    if (_segmentNumRecord > _segmentFlushMaxNumRecords) {
+    if (_segmentNumRecord >= _segmentFlushMaxNumRecords) {
       flush();
     }
   }
 
-  @Override
-  public void snapshotState(FunctionSnapshotContext functionSnapshotContext)
-      throws Exception {
-    throw new UnsupportedOperationException("snapshotState is invoked in Pinot sink");
-  }
-
-  @Override
-  public void initializeState(FunctionInitializationContext functionInitializationContext)
-      throws Exception {
-    // no initialization needed
-    // ...
-  }
-
   private void flush()
       throws Exception {
     URI segmentURI = _segmentWriter.flush();
     LOG.info("Pinot segment writer flushed with {} records to {}", _segmentNumRecord, segmentURI);
     _segmentNumRecord = 0;
+    _pendingRows.clear();
```

Comment:
should clear after the upload was successful?

Reply:
I tried this at first, but upload appears to happen asynchronously and the segmentWriter has already been cleared before the upload.

Reply:
in that case we should definitely call `_pendingRows.clear()` after the async SegmentUpload job returns a success future right?

Reply:
yes that makes sense